pkg/server/image_pull.go

```diff
@@ -17,14 +17,226 @@ limitations under the License.
+				RepoDigests: []string{digest},
+				Size:        uint64(size),          // TODO(mikebrow):  compressed or uncompressed size? using compressed
+				Auth:        r.GetAuth(),           // TODO(mikebrow): what to store here
+				CreatedAt:   time.Now().UnixNano(), // TODO(mikebrow): should this be the image created at time?
+			}
+			if err = c.imageMetadataStore.Create(*meta); err != nil {
+				return &runtime.PullImageResponse{ImageRef: digest},
+					fmt.Errorf("pulled image `%q` but failed to store metadata for digest: %s err: %v", image, digest, err)
+			}
+		}
+	}
+
+	// Return the image digest
+	return &runtime.PullImageResponse{ImageRef: digest}, err
+}
+
+// imageReferenceResolver attempts to resolve the image reference into a name
+// and manifest via the containerd library call..
+//
+// The argument `ref` should be a scheme-less URI representing the remote.
+// Structurally, it has a host and path. The "host" can be used to directly
+// reference a specific host or be matched against a specific handler.
+//
+// The returned name should be used to identify the referenced entity.
+// Dependending on the remote namespace, this may be immutable or mutable.
+// While the name may differ from ref, it should itself be a valid ref.
+//
+// If the resolution fails, an error will be returned.
+func (c *criContainerdService) imageReferenceResolver(ctx context.Context, ref string) (resolvedImageName string, manifest imagespec.Manifest, compressedSize uint64, err error) {
+	var (
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	// This first go routine in the local error group eg will resolve the image
+	// name; place that in the image store with another go routine; then dispatch
+	// a handler to fetch the object the manifest
+	resolver := docker.NewResolver()
+	eg, egctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		var (
+			egerr   error
+			fetcher remotes.Fetcher
+		)
+
+		resolvedImageName, desc, fetcher, egerr = resolver.Resolve(egctx, ref)
+		if egerr != nil {
+			return egerr
+		}
+
+		eg.Go(func() error {
+			return c.imageStore.Put(egctx, resolvedImageName, desc)
+		})
+
+		return containerd_images.Dispatch(
+			egctx,
+			remotes.FetchHandler(c.contentIngester, fetcher),
+			desc)
+	})
+
+	errs := make(chan error)
+	go func() {
+		defer close(errs)
+		errs <- eg.Wait()
+	}()
+
+	select {
+	case err = <-errs:
+		if err != nil {
+			return resolvedImageName, manifest, compressedSize, err
+		}
+	case <-egctx.Done():
+	}
+
+	image, err := c.imageStore.Get(ctx, resolvedImageName)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("get failed for image:%q err: %v", resolvedImageName, err)
+	}
+	p, err := content.ReadBlob(ctx, c.contentProvider, image.Target.Digest)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize,
+			fmt.Errorf("readblob failed for digest:%q err: %v", image.Target.Digest, err)
+	}
+	err = json.Unmarshal(p, &manifest)
```

Comment:
@mikebrow [manifest.Config](https://github.com/opencontainers/image-spec/blob/master/specs-go/v1/manifest.go#L25) is different from [image.Config](https://github.com/opencontainers/image-spec/blob/master/specs-go/v1/config.go#L24).